Implement Observable#toArray

maraisr opened the pull request:
#### 59e8dc1704c4ed34fb6397f471553913aaca3c4d
<pre>
Implement Observable#toArray
<a href="https://bugs.webkit.org/show_bug.cgi?id=277514">https://bugs.webkit.org/show_bug.cgi?id=277514</a>

Reviewed by NOBODY (OOPS!).

This change introduces the `.toArray` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.

Spec: &lt;<a href="https://wicg.github.io/observable/#dom-observable-toarray">https://wicg.github.io/observable/#dom-observable-toarray</a>&gt;

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.worker-expected.txt: Rebaseline.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserver.h:
(WebCore::InternalObserver::globalVM const): Added helper function, as
used often.
(WebCore::InternalObserver::InternalObserver):
* Source/WebCore/dom/InternalObserverToArray.cpp: Added.
(WebCore::createInternalObserverOperatorToArray):
* Source/WebCore/dom/InternalObserverToArray.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::toArray):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes `Promise&lt;sequence&lt;any&gt;&gt; toArray(options)` as a promise-returning operator.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59e8dc1704c4ed34fb6397f471553913aaca3c4d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32395 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118862 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114625 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40958 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85739 "Found 1 new test failure: imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36373 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115610 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101349 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66047 "Found 144 new API test failures: /WPE/TestLoaderClient:/webkit/WebKitWebView/loading-status, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification, /WPE/TestWebKitFindController:/webkit/WebKitFindController/options, /WPE/TestLoaderClient:/webkit/WebKitWebView/reload, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/script-message-received, /WPE/TestWebKitWebView:/webkit/WebKitWebView/run-javascript, /WPE/TestWebKitWebView:/webkit/WebKitWebView/disable-web-security, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification-initial-permission-allowed, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme, /WPE/TestSSL:/webkit/WebKitWebView/tls-subresource ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25665 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19485 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62621 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122083 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94608 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97586 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94350 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39465 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17275 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35828 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39625 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45113 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39264 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41003 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on Source/WebCore/dom/InternalObserverToArray.cpp:
```diff
@@ -0,0 +1,102 @@
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+        m_list.append({ globalObject->vm(), value });
```
To the reviewer; I'd like to draw your attention here. When we collect the `value`'s into this vector, with that being a `Strong<JSC::Unknown>`—are the values kept alive indefinitely? Or only up until the `Vector` itself is destructed?

Wondering if there is a way to only keep the values alive for as long as this operator is kept alive for. As to say, is this code "right" with respects to GC/memory.

cdumez replied:
> To the reviewer; I'd like to draw your attention here. When we collect the value's into this vector, with that being a Strong<JSC::Unknown>—are the values kept alive indefinitely? Or only up until the Vector itself is destructed?

Until the vector is destroyed.

maraisr replied:
Oh perfect, so works as we'd expect. Thanks for clarification

cdumez commented on Source/WebCore/dom/InternalObserverToArray.cpp:
```diff
@@ -0,0 +1,102 @@
+    {
+    }
+
+    Vector<JSC::Strong<JSC::Unknown>> m_list;
```
Personally, I'd feel better if we didn't use `JSC::Strong` for data members. It is too easy to cause leaks. Could we use JSValueInWrappedObject inside the Vector and visit the values inside the `visitAdditionalChildren(JSC::AbstractSlotVisitor&)` member function in this class? We can convert to a Vector of JSC::Strong in `complete()`

maraisr replied:
Would we loop every member of the vector in visitAdditonalChildren? And if so are we concerned with say a vector with thousands of members?

Would be nice if there was something like a VectorWrapped

cdumez replied:
I don't know if there is a more efficient data type. @MenloDorian or @Constellation may have a proposal.

maraisr replied:
I don't think we can use JSValueInWrappedObject here, since the vector will want to grow in size—and move its contents. And JSValueInWrappedObject doesnt allow for moving.

```
error: call to deleted constructor of 'WebCore::JSValueInWrappedObject'
  139 |             new (NotNull, dst) T(WTFMove(*src));

note: in instantiation of member function 'WTF::VectorMover<false, WebCore::JSValueInWrappedObject>::move' requested
      here
  285 |         VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(src, srcEnd, dst);
```

darinadler replied:
The alternative to strong references is to find some other way make sure the objects are marked during garbage collection.

One way to do this would be to have a single `JSValueInWrappedObject` and store the items in the list in a JavaScript array stored in that, using functions like `putDirectIndex`.

annevk commented on Source/WebCore/dom/InternalObserverToArray.cpp:
```diff
@@ -0,0 +1,102 @@
+    {
+    }
+
+    Ref<DeferredPromise> protectedPromise() const { return m_promise; }
```
Since m_promise is const, this isn't needed.

maraisr replied:
Has something changed in the last 5 months, I was given feedback in previous operators' PRs to have these protected things. But can see youve been removing them, so all good! Will push an update to remove it, in anycase. One moment. 

Also, sorry I dropped the ball in getting these PR's out... :/ Would be awesome if we can pair on wrapping up this one as well (https://github.com/WebKit/WebKit/pull/37757)

annevk replied:
This might have changed in the last five months, I'm not sure, see https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#mark-data-members-that-are-smart-pointers-as-const-if-they-never-get-reassigned

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt:
```diff
@@ -1,8 +1,9 @@
+PASS toArray(): complete() resolves promise; subsequent error()s report the exceptions
+PASS toArray(): Subscribing with an aborted signal returns an immediately rejected promise
+PASS toArray(): Aborting the passed-in signal rejects the returned promise
+PASS Operator Promise abort ordering
```
missing a newline here according to bots

darinadler commented on Source/WebCore/dom/InternalObserverToArray.cpp:
```diff
@@ -0,0 +1,100 @@
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
```
Seems inelegant to have this all written out like this. Maybe we should package it all into a helper function? Especially if it's a recurring pattern seen elsewhere.

maraisr replied:
Oh yeah totally agree, remember doing this a bunch for the other operators. 2 things;
1. Are you okay for me to land a helper function + integration in a separate diff
2. Where would you recommend this helper method go? Or are you simply meaning inline to this file?

darinadler replied:
I think I’d start by putting a `protected` helper function in the `InternalObserver` class, maybe named `globalVM`. After landing this change, there would now be 6 places where that would be used.

darinadler commented on Source/WebCore/dom/InternalObserverToArray.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+    }
+
+    Vector<JSC::Strong<JSC::Unknown>> m_list;
```
Let me try to be even more explicit than @cdumez. I think this may be incorrect and will create a permanent leak if there are reference cycles. We should create a test that proves we can collect even if you try to make a reference cycle that intentionally covers the most difficult edge case; it's possible we can't use `Strong` and pass such a test.

maraisr replied:
Great call out!! 🤩 Ill have to figure out where and how to write that test, but keen as!! Hypothetically (maybe even more) what would an alternate approach be should the test prove this leaks?

darinadler replied:
An alternate approach needs to put the items where they will be marked during garbage collection as a side effect of marking the `InternalObserverToArray` object. I don’t know offhand all the details of how that would be done are; it can be tricky to get right because of the garbage collection happening on another thread.

darinadler replied:
One thing that might work would be using a temporary JavaScript array object to hold the list, and then have a reference to that JavaScript array be stored in a `JSValueInWrappedObject`.

maraisr committed:
Implement Observable#toArray
https://bugs.webkit.org/show_bug.cgi?id=277514

Reviewed by NOBODY (OOPS!).

This change introduces the `.toArray` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.

Spec: <https://wicg.github.io/observable/#dom-observable-toarray>

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.worker-expected.txt: Rebaseline.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserver.h:
(WebCore::InternalObserver::globalVM const): Added helper function, as
used often.
(WebCore::InternalObserver::InternalObserver):
* Source/WebCore/dom/InternalObserverToArray.cpp: Added.
(WebCore::createInternalObserverOperatorToArray):
* Source/WebCore/dom/InternalObserverToArray.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::toArray):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes `Promise<sequence<any>> toArray(options)` as a promise-returning operator.
